framework/filters/AccessRule.php

```diff
@@ -161,8 +191,13 @@ protected function matchRole($user)
                 if (!$user->getIsGuest()) {
                     return true;
                 }
-            } elseif ($user->can($role)) {
-                return true;
+            } else {
+                if (!isset($roleParams)) {
```

Comment:
> or anyone can upload images anonymously ('?') but only the creator ip can delete them?

the if (isset(...)) ensures that it is evaluated when needed first, but then not evaluated multiple times.